pkg/ccl/changefeedccl/kvevent/blocking_buffer.go

```diff
@@ -210,39 +237,54 @@ var bufferEntryPool = sync.Pool{
 	},
 }
 
-var _ quotapool.Request = (*bufferEntry)(nil)
+func newBufferEntry(e Event, alloc int64) *bufferEntry {
+	be := bufferEntryPool.Get().(*bufferEntry)
+	be.e = e
+	be.alloc = alloc
+	be.next = nil
+	return be
+}
+
+// memRequest is a quotapool memory request.
+type memRequest int64
+
+var _ quotapool.Request = memRequest(0)
 
 // Acquire implements quotapool.Request interface.
-func (r *bufferEntry) Acquire(
+func (r memRequest) Acquire(
 	ctx context.Context, resource quotapool.Resource,
 ) (fulfilled bool, tryAgainAfter time.Duration) {
-	res := resource.(*blockingBufferQuotaPool)
-	if res.canAllocateBelow > 0 {
-		if res.allocated > res.canAllocateBelow {
+	quota := resource.(*memQuota)
+	if quota.canAllocateBelow > 0 {
+		if quota.allocated > quota.canAllocateBelow {
 			return false, 0
 		}
-		res.canAllocateBelow = 0
+		quota.canAllocateBelow = 0
 	}
-	if err := res.acc.Grow(ctx, r.alloc); err != nil {
-		if res.allocated == 0 {
-			// We've failed but there's nothing outstanding, that means we're doomed
-			// to fail forever and should propagate the error.
-			r.err = err
-			return true, 0
+
+	if err := quota.acc.Grow(ctx, int64(r)); err != nil {
+		if quota.allocated == 0 {
+			// We've failed but there's nothing outstanding.  It seems that this request
+			// is doomed to fail forever. However, that's not the case since our memory
+			// quota is tied to a larger memory pool.  We failed to allocate memory for this
+			// single request, but we may succeed if we try again later since some other
+			// process may release it into the pool.
+			// TODO(yevgeniy): Consider making retry configurable; possibly with backoff.
+			return true, time.Second
```

Comment:
This seems scary, we may never succeed. This would be a very hard thing to debug.